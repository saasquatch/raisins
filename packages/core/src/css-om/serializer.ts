import csstree from "css-tree";

/**
 * Converts a plain CSS node AST into a css string
 *
 * @param cssNode
 * @returns
 */
export function serializer(cssNode: csstree.CssNodePlain) {
  // Uses the extra plain clone to ensure `css-tree` doesn't mutate the static objects
  const clone = JSON.parse(JSON.stringify(cssNode));
  return csstree.generate(csstree.fromPlainObject(clone), {});
}

export default serializer;
