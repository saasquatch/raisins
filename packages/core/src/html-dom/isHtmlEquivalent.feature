Feature: HTML Equivalency Checker

	HTML Comparison is hard because there are many ways of writing HTML where it might be a different string, but equivalent for most purposes.
	For example, minified HTML should be equivalent to the source HTML, even though it has less whitespace.

	Based roughty on `node.isEqualNode` (https://developer.mozilla.org/en-US/docs/Web/API/Node/isEqualNode) except that it's not implemented in jsdom,
	and it doesn't look for CSS equivalency.

	@motivating
	Scenario Outline: Differences in simple nodes will fail
		Ensures inequvalent nodes should fail when compared.

		Given one HTML string <First>
		And another HTML string <Second>
		Then they will not be equivalent

		Examples:
			| First                    | Second                |
			| <div></div>              |                       |
			| <div>                    | </div>                |
			| <div></div>              | <span></span>         |
			| <div id></div>           | <div></div>           |
			| <div id=""></div>        | <div></div>           |
			| <div id="value"></div>   | <div></div>           |
			| <div>Has text</div>      | <div></div>           |
			| <div><span></span></div> | <div></div>           |
			| <!DOCTYPE html>          | <!--DOCTYPE html-->   |
			| <!DOCTYPE html>          | <?DOCTYPE html?>      |
			| <?DOCTYPE html?>         | <!--DOCTYPE html-->   |
			| <div style></div>        | <div></div>           |
			| <div style=""></div>     | <div></div>           |
			| <div style=style></div>  | <div></div>           |
			| <noscript></noscript>    | <textarea></textarea> |

	@motivating
	Scenario Outline: Optional tags can be omitted
		Spec: https://html.spec.whatwg.org/multipage/syntax.html#optional-tags

		Given one HTML string <First>
		And another HTML string <Second>
		Then they will be equivalent

		Examples:
			| First                                     | Second                                                                           |
			| <div></div>                               | <html><head></head><body><div></div></body><html>                                |
			| <body><div></div></body>                  | <html><head></head><body><div></div></body><html>                                |
			| <html><div></div></html>                  | <html><head></head><body><div></div></body><html>                                |
			| <meta charset="UTF-8"><div></div>         | <html><head><meta charset="UTF-8"></head><body><div></div></body<html>           |
			| <title>hello world</title><p>text ...</p> | <html><head><title>hello world</title></head><body><p>text ...</p></body></html> |
			| <html><!--comment--><div></div></html>    | <html><!--comment--><head></head><body><div></div></body<html>                   |
			| <head><!--comment--></head><div></div>    | <html><head><!--comment--></head><body><div></div></body<html>                   |
			| <body><!--comment--><div></div></body>    | <html><head></head><body><!--comment--><div></div></body><html>                  |
			| <html lang="en"><div></div></html>        | <html lang="en"><head></head><body><div></div></body<html>                       |
			| <!--comment--><div></div>                 | <!--comment--><html><head></head><body><div></div></body<html>                   |
			| <!DOCTYPE html><div></div>                | <!DOCTYPE html><html><head></head><body><div></div></body><html>                 |


	@motivating
	Scenario Outline: Optional tags cannot be omitted if immediately folllowed by a comment or contain any attributes
		Spec: https://html.spec.whatwg.org/multipage/syntax.html#optional-tags

		Given one HTML string <First>
		And another HTML string <Second>
		Then they will not be equivalent

		Examples:
			| First                                     | Second                                                                                     |
			| <html><!--comment--><div></div></html>    | <!--comment--><html><head></head><body><div></div></body<html>                             |
			| <head><!--comment--></head><div></div>    | <!--comment--><html><head></head><body><div></div></body<html>                             |
			| <body><!--comment--><div></div></body>    | <!--comment--><html><head></head><body><div></div></body<html>                             |
			| <html lang="en"><div></div></html>        | <html><head></head><body><div></div></body<html>                                           |
			| <title>hello world</title><p>text ...</p> | <html lang="en"><head><title>hello world</title></head><body><p>text ...</p></body></html> |
			| <div></div>                               | <!DOCTYPE html><html><head></head><body><div></div></body><html>                           |

	@motivating
	Scenario Outline: Misnested tags are handled as per specifications
		Spec: https://html.spec.whatwg.org/multipage/parsing.html#misnested-tags:-b-i-/b-/i

		Given one HTML string <First>
		And another HTML string <Second>
		Then they will be equivalent

		Examples:
			| First                      | Second                            |
			| <p>1<b>2<i>3</b>4</i>5</p> | <p>1<b>2<i>3</i></b><i>4</i>5</p> |
			| <b>1<p>2</b>3</p>          | <b>1</b><p><b>2</b>3</p>          |


	@motivating
	Scenario Outline: Attribute order does not matter
		Spec: https://dom.spec.whatwg.org/#ref-for-dom-node-isequalnode%E2%91%A0

		Given one HTML string <First>
		And another HTML string <Second>
		Then they will be equivalent

		Examples:
			| First                      | Second                     |
			| <div a b></div>            | <div b a></div>            |
			| <span a="value" b></span>  | <span b a="value"></span>  |
			| <div><h1 a b></h1></div>   | <div><h1 b a></h1></div>   |
			| <div a b><p b a></p></div> | <div b a><p a b></p></div> |

	@motivating
	Scenario Outline: Boolean attirbutes
		Spec: https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes

		Given one HTML string <First>
		And another HTML string <Second>
		Then they will be equivalent

		Examples:
			| First                        | Second                       |
			| <details open></details>     | <details open=""></details>  |
			| <details open=""></details>  | <details open=""></details>  |
			| <details style></details>    | <details style=""></details> |
			| <details style=""></details> | <details style=""></details> |

	@motivating
	Scenario Outline: HTML singleton tags with no closing bracket
		Closing tag should not be needed in certain html tags (e.g. <img>)
		Spec: https://html.spec.whatwg.org/multipage/syntax.html#void-elements

		Given one HTML string <First>
		And another HTML string <Second>
		Then they will be equivalent

		Examples:
			| First                  | Second                       |
			| <img>                  | <img></img>                  |
			| <area>                 | <area></area>                |
			| <base>                 | <base></base>                |
			| <input>                | <input></input>              |
			| <embed>                | <embed></embed>              |
			| <meta>                 | <meta></meta>                |
			| <hr>                   | <hr />                       |
			| <col>                  | <col />                      |
			| <link>                 | <link />                     |
			| <source>               | <source />                   |
			| <track>                | <track />                    |
			| <wbr>                  | <wbr />                      |
			| <br>                   | </br>                        |
			| <br>                   | <br/>                        |
			| <img src="helloworld"> | <img src="helloworld"></img> |

	@motivating
	Scenario Outline: Extra whitespace does not matter
		Given one HTML string <First>
		And another HTML string <Second>
		Then they will be equivalent

		Examples:
			| First                     | Second                          |
			| <div> </div>              | <div>  </div>                   |
			| <div     open   ></div>   | <div open></div>                |
			| <body><div> </div></body> | <body><div>       </div></body> |

	@motivating
	Scenario Outline: Whitespace does matter
		The presences of whitespace matters in certain situations because it affects text rendering.
		Spec:: https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Whitespace

		Given one HTML string <First>
		And another HTML string <Second>
		Then they will not be equivalent

		Examples:
			| First                                 | Second                               |
			| <div>Hello <b>World</b></div>         | <div>Hello<b>World</b></div>         |
			| <span>Hello <span>World</span></span> | <span>Hello<span>World</span></span> |

	@motivating
	Scenario Outline: Encoded characters do not matter

		Given one HTML string <First>
		And another HTML string <Second>
		Then they will be equivalent

		Examples:
			| First                                                                 | Second                                                                                             |
			| <a name="&lpos=Class&lid=<span style='color: green'>CT</span>"  ></a> | <a  name="&amp;lpos=Class&amp;lid=&lt;span style=&apos;color: green&apos;&gt;CT&lt;/span&gt;"></a> |
			| <img src="" alt="(Photo: NASA&amp;JAXA/Inode)>                        | <img src="" alt="(Photo: NASA&JAXA/Inode)>                                                         |
			| <div>And &#x2014; poof &#x2014; time is gone.</div>                   | <div>And &mdash; poof &mdash; time is gone.</div>                                                  |

	@motivating
	Scenario Outline: Unequivalent CSS styles will cause failures
		Given one HTML string <First>
		And another HTML string <Second>
		Then they will not be equivalent

		Examples:
			| First                                          | Second                                     |
			| <div style="display: none; "></div>            | <div style="bold:true;"></div>             |
			| <div style="display: blue; "></div>            | <div style="bold: contents;"></div>        |
			| <div style='opacity: 0.5"></div>               | <div style="opacity: 0.5"></div>           |
			| <style> body {border: 1px solid red;} </style> | <style> body {color: blue;} </style>       |
			| <style> body {display: none;} </style>         | <style> body {DISPLAY : CONTENTS} </style> |
			| <style> body {color: red;} </style>            | <style> body {color: red}; </style>        |

	@motivating
	Scenario Outline: Unequivalent CSS styles will ignore text differences
		Given one HTML string <First>
		And another HTML string <Second>
		Then they will be equivalent

		Examples:
			| First                                           | Second                                                        |
			| <div style="display: none;"></div>              | <div style="display:none"></div>                              |
			| <div style="display: none;"></div>              | <div style="DISPLAY: NONE;"></div>                            |
			| <div style="   display: none; "></div>          | <div style="display: none;"></div>                            |
			| <div style="opacity: 0.5"></div>                | <div style='opacity: 0.5'></div>                              |
			| <style>     body {   color:   red} </style>     | <style> body {color: red;} </style>                           |
			| <style type="text/css">body{color:red;}</style> | <style    type='text/css'  >  body  {color:   red}   </style> |
			| <style> body {display: none;} </style>          | <style> body {DISPLAY : NONE} </style>                        |
			| <style> body {/* color: red; */} </style>       | <style> body {/* color: blue; */} </style>                    |
			| <style> body {} </style>                        | <style> body {/* color: blue; */} </style>                    |

	@motivating
	Scenario Outline: Ignore comments flag will ignore comments in comparison

		Given one HTML string <First>
		And another HTML string <Second>
		And ignoreComments option is true
		Then they will be equivalent

		Examples:
			| First                                | Second                               |
			| <div>hello world</div>               | <!--comment--><div>hello world</div> |
			| <!--a--><div>hello world</div>       | <!--b--><div>hello world</div>       |
			| <div>hello world</div><!--comment--> | <div>hello world</div>               |