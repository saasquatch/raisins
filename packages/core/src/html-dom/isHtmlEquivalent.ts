import * as css5 from "css";
import * as parse5 from "parse5";
import expect from "expect";

/**
 * Validates CSS
 */
interface css5Options {
  inline?: boolean;
}

function parseCSS(input: string, options?: css5Options): string {
  input = input.toLowerCase();
  if (options?.inline) input = "*{" + input + "}";
  try {
    input = css5.stringify(css5.parse(input), {
      compress: true
    });
    input = css5.stringify(css5.parse(input), {
      compress: true
    });
  } catch {
    return "";
  }
  input = input.replace(/\s+/g, "");
  if (options?.inline) input = input.slice(2, -1);
  return input;
}

/**
 * Cleans css style values from html
 */
function cleanUpCSS(html: string) {
  html = html
    .replace(/(?<=style=(["']))(?:(?=(\\?))\2.)*?(?=\1)/gs, input =>
      parseCSS(input, { inline: true })
    )
    .replace(
      /(<style.*?>)(.*?)(<\/style>)/gs,
      (_, start, input, end) => start + parseCSS(input) + end
    );
  return html;
}

/**
 * Compares two HTML strings to see if they are equivalent
 *
 * @param expected original html string
 * @param received comparison html string
 * @returns
 */
export function isHtmlEquivalent(
  expected: string,
  received: string,
  purifyOptions?: PurifyOptions,
  parseOptions?: parse5.ParserOptions
): true | never {
  /*
   * Initial pass through parse5
   */
  expected = parse5.serialize(parse5.parse(expected, parseOptions));
  received = parse5.serialize(parse5.parse(received, parseOptions));

  /*
   * Data massage on html
   */
  expected = cleanUpCSS(expected);
  received = cleanUpCSS(received);

  /*
   * Second trip to compare
   */
  const parsedExpected = parse5.parse(expected, parseOptions);
  const parsedReceived = parse5.parse(received, parseOptions);

  /*
   * Data massage on parse5 nodes
   */
  purify(parsedExpected, purifyOptions);
  purify(parsedReceived, purifyOptions);

  /*
   * Test node equivalence with deep comparison
   */
  expect(parsedReceived).toStrictEqual(parsedExpected);

  return true;
}
interface PurifyOptions {
  ignoreComments?: boolean;
}

/*
  Absolves parse5 nodes from their sins by:
    - removing parentNode circular references
    - sorting attributes to be in the same order (if they exist)
    - removing text nodes caused by whitespaces between elements
      or containing html elements
*/
function purify(node: any, options?: PurifyOptions | undefined) {
  for (var i = 0; i < node.childNodes.length; i++) {
    var child = node.childNodes[i];
    if (child.hasOwnProperty("parentNode")) {
      delete child["parentNode"];
    }
    if (child.hasOwnProperty("attrs")) {
      child.attrs.sort(attributeCompare);
    }
    if (child.nodeName === "#text" && !/\S/.test(child.value)) {
      node.childNodes.splice(i, 1);
      i--;
    }
    if (options?.ignoreComments && child.nodeName === "#comment") {
      node.childNodes.splice(i, 1);
      i--;
    }
    if (child.hasOwnProperty("childNodes")) {
      purify(child, options);
    }
  }
}

/*
  Sort function for comparing two nodes by name
 */
function attributeCompare(a: any, b: any) {
  if (a.name < b.name) {
    return -1;
  }
  if (a.name > b.name) {
    return 1;
  }
  return 0;
}

export default isHtmlEquivalent;
