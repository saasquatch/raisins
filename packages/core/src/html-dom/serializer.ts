import { CssNodePlain } from "css-tree";
import { encodeXML, escapeAttribute } from "entities";
import cssSerializer from "../css-om/serializer";
import foreignNames from "./foreignNames";
// eslint-disable-next-line prettier/prettier
import type {
  RaisinCommentNode,
  RaisinElementNode,
  RaisinNode,
  RaisinNodeWithChildren,
  RaisinProcessingInstructionNode,
  RaisinTextNode
} from "./RaisinNode";
import { getParents, visit } from "./util";

/**
 *
 *  Forked from: https://github.com/cheeriojs/dom-serializer
 *
 *
 * LV: I don't like forking libraries, but in the case of DOM-Handler the programming style is very mutable. Even in forking the
 * code I found that the serializer was **mutating tag names as it was serializing them**. This feels like it would have unintended
 * side effects throughout the state tree.
 *
 * An alternate approach was considered for first turning `RaisinNode` nodes back into `domhandler` nodes, but that has previously
 * proven to be a cumbersome process:
 *  - `domhandler` nodes have multiple names and copies of the same property (e.g. `name` and `tagName`)
 *  - `domhandler` nodes need to be linked up with a `parent` and `prev` and `next`, even though the serializer only relies on `parent`
 *  - `domhandler` constructors for creating instances of their classes have caused build errors in the Raisin.js repo
 *
 *
 * This also adds a few features:
 *  - `style` tags have their CSS contents parsed and serialized
 *
 *  TODO: Also fork the test suite
 *
 */

const { attributeNames, elementNames } = foreignNames;
export interface DomSerializerOptions {
  xmlMode?: boolean | "foreign";
  emptyAttrs?: boolean;
  selfClosingTags?: boolean;
}

// Type for parents
type Parents = WeakMap<RaisinNode, RaisinNodeWithChildren>;

const unencodedElements = new Set([
  "style",
  "script",
  "xmp",
  "iframe",
  "noembed",
  "noframes",
  "plaintext",
  "noscript",
]);

/**
 * Format attributes
 */
function formatAttributes(
  attributes: Record<string, string | null> | undefined,
  opts: DomSerializerOptions
) {
  if (!attributes) return;

  return Object.keys(attributes)
    .map((key) => {
      const value = attributes[key] ?? "";

      if (opts.xmlMode === "foreign") {
        /* Fix up mixed-case attribute names */
        key = (attributeNames[key as keyof typeof attributeNames]) ?? key;
      }

      if (!opts.emptyAttrs && !opts.xmlMode && value === "") {
        // For boolean attributes
        // See: https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes
        return key;
      }
      
      /**
       * Encodes all characters that have to be escaped in HTML attributes,
       * following {@link https://html.spec.whatwg.org/multipage/parsing.html#escapingString}.
       *
       * @param data String to escape.
       */
      return `${key}="${escapeAttribute(value)}"`;
    })
    .join(" ");
}

/**
 * Self-enclosing tags
 */
const singleTag = new Set([
  "area",
  "base",
  "basefont",
  "br",
  "col",
  "command",
  "embed",
  "frame",
  "hr",
  "img",
  "input",
  "isindex",
  "keygen",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr",
]);

/**
 * Renders a DOM node or an array of DOM nodes to a string.
 *
 * Can be thought of as the equivalent of the `outerHTML` of the passed node(s).
 *
 * @param node Node to be rendered.
 * @param options Changes serialization behavior
 */
export default function serializer(
  node: RaisinNode | RaisinNode[],
  options: DomSerializerOptions = {}
): string {
  const nodes: RaisinNode[] = Array.isArray(node)
    ? (node as RaisinNode[])
    : [node];

  let output = "";

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    const parents = getParents(node);
    output += renderNode(node, options, parents);
  }

  return output;
}

function renderNode(
  node: RaisinNode,
  options: DomSerializerOptions,
  parents: Parents
): string {
  return visit(node, {
    onText: (n) => renderText(n, options, parents),
    onElement: (n, children) => renderTag(n, options, parents, children),
    onRoot(_, children?: string[]) {
      return children?.join("");
    },
    onStyle(n) {
      const cssContents = n.contents ? cssSerializer(n.contents) : "";
      const attribs = formatAttributes(n.attribs, options);
      return `<style${attribs ? " " + attribs : ""}>${cssContents}</style>`;
    },
    onDirective: renderDirective,
    onComment: renderComment,
  }) ?? "";
}

const foreignModeIntegrationPoints = new Set([
  "mi",
  "mo",
  "mn",
  "ms",
  "mtext",
  "annotation-xml",
  "foreignObject",
  "desc",
  "title",
]);

const foreignElements = new Set(["svg", "math"]);

function renderTag(
  elem: RaisinElementNode,
  opts: DomSerializerOptions,
  parents: Parents,
  children?: string[]
) {
  const parent = parents.get(elem);
  let { tagName } = elem;
  // Handle SVG / MathML in HTML
  if (opts.xmlMode === "foreign") {
    /* Fix up mixed-case element names */
    tagName = elementNames[elem.tagName as keyof typeof elementNames] ?? elem.tagName;
    /* Exit foreign mode at integration points */
    if (
      parent &&
      foreignModeIntegrationPoints.has((parent as RaisinElementNode).tagName)
    ) {
      opts = { ...opts, xmlMode: false };
    }
  }
  if (!opts.xmlMode && foreignElements.has(tagName)) {
    opts = { ...opts, xmlMode: "foreign" };
  }

  let tag = `<${tagName}`;
  const attribs = formatAttributes(elem.attribs, opts);

  if (attribs) {
    tag += ` ${attribs}`;
  }

  let style = elem.style && formatStyle(elem.style);
  if (style) {
    tag += ` ${style}`;
  }

  if (
    elem.children.length === 0 &&
    (opts.xmlMode
      ? // In XML mode or foreign mode, and user hasn't explicitly turned off self-closing tags
        opts.selfClosingTags !== false
      : // User explicitly asked for self-closing tags, even in HTML mode
        opts.selfClosingTags && singleTag.has(tagName))
  ) {
    if (!opts.xmlMode) tag += " ";
    tag += "/>";
  } else {
    tag += ">";
    if (elem.children.length > 0) {
      tag += children?.join("") ?? "";
    }

    if (opts.xmlMode || !singleTag.has(tagName)) {
      tag += `</${tagName}>`;
    }
  }

  return tag;
}

function renderDirective(elem: RaisinProcessingInstructionNode) {
  return `<${elem.data}>`;
}

function renderText(
  elem: RaisinTextNode,
  opts: DomSerializerOptions,
  parents: Parents
) {
  let data = elem.data || "";
  const parent = parents.get(elem);
  // If entities weren't decoded, no need to encode them back
  if (
    !(
      !opts.xmlMode &&
      parent &&
      unencodedElements.has((parent as RaisinElementNode).tagName)
    )
  ) {
    data = encodeXML(data);
  }

  return data;
}

function renderComment(elem: RaisinCommentNode) {
  return `<!--${elem.data}-->`;
}

function formatStyle(style: CssNodePlain): string | undefined {
  if (!style) return;
  const content = cssSerializer(style);
  return `style="${content.replace(/"/g, "&quot;")}"`;
}