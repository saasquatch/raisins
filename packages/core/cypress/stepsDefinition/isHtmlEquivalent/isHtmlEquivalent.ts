import { And, Given, Then } from "cypress-cucumber-preprocessor/steps";
import expect from "expect";
import { isHtmlEquivalent } from "../../../src/html-dom/isHtmlEquivalent";

var html_a: string;
var html_b: string;
var ignoreComments: boolean = false;

Given(/^one HTML string (.*)$/, (html: string) => {
  html_a = html;
});

And(/^another HTML string (.*)$/, (html: string) => {
  html_b = html;
});

And(/^ignoreComments option is true$/, () => {
  ignoreComments = true;
});

Then(/^they will be equivalent$/, () => {
  expect(isHtmlEquivalent(html_a, html_b, { ignoreComments })).toBe(true);
});

Then(/^they will not be equivalent$/, () => {
  expect(() => {
    isHtmlEquivalent(html_a, html_b);
  }).toThrowError();
});
