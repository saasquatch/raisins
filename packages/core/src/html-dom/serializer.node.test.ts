import expect from "expect";
import glob from "glob";
import { fs } from "mz";
import isHtmlEquivalent from "./isHtmlEquivalent";
import parse from "./parser";
import serializer from "./serializer";

describe("Parse + serialize", () => {
  test("Can parse simple HTML", () => {
    const source = `<div>Example</div>`;
    const node = parse(source);
    expect(node).toMatchSnapshot();

    const out = serializer(node);

    expect(out).toBe(source);
  });

  describe("Can parse Email Benchmark cases", () => {
    const files = glob.sync(process.cwd() + "/benchmark-emails/**/*.html", {
      absolute: true
    });
    expect(files?.length).toBeGreaterThan(0);
    for (const file of files) {
      test(`File ${files.indexOf(file)}: ${file}`, async () => {
        var source = await fs.readFile(file, "utf-8");
        const raisinNode = parse(source, {
          cleanWhitespace: false
        });
        const raisinString = serializer(raisinNode);
        isHtmlEquivalent(source, raisinString);
      });
    }
  });

  describe("Can parse HTML Benchmark cases", () => {
    const files = glob.sync(process.cwd() + "/benchmark-files/**/*.html", {
      absolute: true
    });
    expect(files?.length).toBeGreaterThan(0);
    for (const file of files) {
      test(`File ${files.indexOf(file)}: ${file}`, async () => {
        var source = await fs.readFile(file, "utf-8");
        const raisinNode = parse(source, {
          cleanWhitespace: false
        });
        const raisinString = serializer(raisinNode);
        isHtmlEquivalent(source, raisinString);
      });
    }
  });
});
