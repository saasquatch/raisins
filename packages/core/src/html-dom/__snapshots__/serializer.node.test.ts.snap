// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse + serialize Can parse simple HTML 1`] = `
Object {
  "children": Array [
    Object {
      "attribs": Object {},
      "children": Array [
        Object {
          "data": "Example",
          "type": "text",
        },
      ],
      "style": undefined,
      "tagName": "div",
      "type": "tag",
    },
  ],
  "type": "root",
}
`;
